rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,438 @@
+where the application can upgrade the Tx data structures. 
+
+**Negative**
+
+- Nodes may have to go through millions of blocks requiring extra memory and
+computation over the transition phase
+- External clients would most likely need to support both block versions over
+the transition period
+- Bootstrapping nodes and light clients require an extra step for verification.
+- We've cornered off some untouchable fields in the block structure. If we
+really wanted to change these we would need a hard upgrade.
+
+## Status
+
+Proposed
+
+## References
+
+- [Upgrade tooling tracking issue](https://github.com/tendermint/tendermint/issues/5595)
+- [Protocol versions](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-016-protocol-versions.md)
+
+## Appendix A: Upgrade terminology
+
+It's important to define upgrades and classify the distinctions so as to ensure
+that everyone has the same understanding of what it means and that we can use
+the same terminology to speak about the different concepts. This appendix aims
+at achieving this.
+
+An upgrade, in the loosest sense, is a change to the Tendermint codebase. For
+simplicity to users, we group changes together and form releases. Every release
+corresponds to a unique software version. Tendermint follows SemVer which
+has strict rules about versioning and makes a distinction between different
+types of releases.
+
+From the node operators perspective, upgrades just mean, stopping the node,
+changing the binary and starting the node again.
+
+A patch release (the last number in x.x.x), is a backwards compatible bug-fix.
+Nodes should be able to perform this upgrade at any time without any affect to
+the network.
+
+A minor release (the middle number in x.x.x), is also backwards compatible but
+it indicates changes in existing features or new features. One can think of
+performance optimizations as a good example. Nodes should also be able to
+perform this upgrade at any time and nodes with different minor versions should
+have no problem interacting with one another.
+
+We call the above two releases as constituting minor upgrades.
+
+A major release (the first number in x.x.x), is a set of incompatible changes.
+This means that the public api has changed and/or the messages that nodes send
+to one another and persist to disk have changed in a way that is incompatible
+with prior releases. A block protocol change is always a major release.
+
+When this happens nodes can't simply stop and restart when they want but must
+coordinate a restart together. There can't be nodes with different versions on
+the same network. Not only this, but upgraded nodes can't read nor verify data
+structures from a prior major version. Thus, an upgrade for a major release
+has so far meant having to create a new chain.
+
+Soft and hard upgrades both refer to major releases but have very different
+properties.
+
+A soft upgrade like minor upgrades can happen asynchronously across the
```

Comment:
It seems that what is meant with soft upgrade is supporting nodes and clients with multiple versions to be present in the network and operate correctly. Hard fork requires everyone to be on the same version. Is this correct?

Reply:
A soft fork is when the latest *software* version i.e 0.34.0 can support an entire blockchain. For example if going from v6 to v7 is a soft fork then v7 must support the same blockchain that v6 supported whether through multi-version support or migrations. By support, I essentially mean be able to verify all data structures. A hard fork means that the latest version no longer supports the prior blockchain. This requires a new blockchain.

Reply:
And what about previous versions of software (since the last hard-fork)? Will it still expected to work with at least heights before soft-fork they are lagging behind?

Reply:
depends on the implementation. With the migrations, you probably won't be able to operate with a lesser version (even at a lower height) because all the blocks will be updated to the latest version and other peers will be serving these latest blocks. Perhaps if you were peered with other nodes that were on the previous version then they would still be serving those blocks. But once you reached the soft upgrade height you wouldn't be able to go any further.

This last bit would be the same with the multi support version. Nodes on prior versions could still process earlier blocks but would halt when they reached the soft upgrade height

Reply:
I guess something else to consider is that because we're running this off consensus, it means we could also get version-forks if 2/3 of validating power is malicious. This would mean that the network forks with some nodes following version x and others following version y

Reply:
Sure we could have the table in the UPGRADES file or somewhere else. It would also be nice to add this in the `tendermint version` command. 

I was thinking that a group of ABCI, block and p2p versions could all correlate to a spec version and then a software version just maps out which spec versions it covers although I'm not sure this would exactly work